### ISSUE TITLE
What is the prefix:'PCAGML' of font:"PCAGML+SourceHanSerifCN-Regular,16.0"?

### ISSUE THREAD

**Issue by forhonourlx** [jstockwin/py-pdf-parser#100]
Hi @jstockwin ,

I am trying to load 3 similar files which is auto generated from same system, but get 3 different prefix of font.
```
>>> set(e.font for e in doc.elements)
{'OHKPGR+SourceHanSerifCN-Regular,10.5', 'Helvetica,12.0', 'OHKPGR+SourceHanSerifCN-Regular,26.0', 'OHKPGR+SourceHanSerifCN-Regular,9.0', 'OHKPGR+SourceHanSerifCN-Regular,12.0', 'OHKPGR+SourceHanSerifCN-Regular,8.5', 'OHKPGR+SourceHanSerifCN-Regular,16.0', 'OHKPGR+SourceHanSerifCN-Regular,14.0'}

>>> set(e.font for e in doc2.elements)
{'Helvetica,12.0', 'FZVVCB+SourceHanSerifCN-Regular,8.5', 'FZVVCB+SourceHanSerifCN-Regular,12.0', 'FZVVCB+SourceHanSerifCN-Regular,26.0', 'FZVVCB+SourceHanSerifCN-Regular,14.0', 'FZVVCB+SourceHanSerifCN-Regular,10.5', 'FZVVCB+SourceHanSerifCN-Regular,16.0', 'FZVVCB+SourceHanSerifCN-Regular,9.0'}

>>> set(e.font for e in doc3.elements)
{'PCAGML+SourceHanSerifCN-Regular,16.0', 'PCAGML+SourceHanSerifCN-Regular,14.0', 'PCAGML+SourceHanSerifCN-Regular,26.0', 'PCAGML+SourceHanSerifCN-Regular,12.0', 'PCAGML+SourceHanSerifCN-Regular,8.5', 'PCAGML+SourceHanSerifCN-Regular,9.0', 'Helvetica,12.0', 'PCAGML+SourceHanSerifCN-Regular,10.5'}
```

What is the prefix:'PCAGML' of font:"PCAGML+SourceHanSerifCN-Regular,16.0"?
And how to use FONT_MAPPING over these diff?

Thanks in advance.

**Comment by jstockwin** [comment]
Hi @forhonourlx,

Thanks for getting in touch.

> What is the prefix...

I have also seen these in certain PDFs, and to be honest I've never looked into why they happen. The fonts are taken from how [pdfminer.six](https://github.com/pdfminer/pdfminer.six) does its extraction. I suspect it's just something to do with how the fonts are given in the PDF, but perhaps there is a bug somewhere and they shouldn't be included.

> And how to use FONT_MAPPING over these diff?

In general, the `font_mapping` argument to the PDFDocument is a dictionary mapping the actual font names to more useful names. I find it helpful to map the fonts to something more meaningful in the document, so for example you might pass `font_mapping = {"OHKPGR+SourceHanSerifCN-Regular,26.0": "heading"}` if that font is usually used for headings.

To get around the prefix issue, you can also use regular expressions as the dictionary keys. In this case,  you also need to pass `font_mapping_is_regex=True`. This will allow you to do something like `font_mapping={"SourceHanSerifCN-Regular,26.0$": "heading"}, font_mapping_is_regex=True`, or perhaps `font_mapping={".*SourceHanSerifCN-Regular,26.0": "heading"}, font_mapping_is_regex=True` (I'm not sure if we force the start and end of the string or not, you'd have to experiment). You can also pass `regex_flags` if you want any flags to be used, for example you can pass `regex_flags=re.IGNORECASE`.

You can see all of these arguments in the [reference documentation for PDFDocument](https://py-pdf-parser.readthedocs.io/en/latest/reference/components.html#py_pdf_parser.components.PDFDocument). Note that although the arguments are for `PDFDocument`, you're probably using `load` or `load_file` to load the document. These arguments can also be passed to the loader, which will pass them through to the `PDFDocument`.

I hope that helps! Please do let me know how you find the tool - It's fairly new and I am keen to get it as user-friendly as possible.

**Comment by forhonourlx** [comment]
Hi @jstockwin ,

Thank you for your sharing the project again, py-pdf-parser is rather perfect. It seems that `font_mapping_is_regex` has already solved the problem.

I came to know this project from a diligent Chinese colleage computer science bloger @爱可可-爱生活 on Weibo.
I also wonder how he found your project, maybe @fly51fly  on github could tell you.

My story is  facing a pdf parser task on my job, coming with a lot of nosql data(also tables) represented in many sections, I search the keywords: 'pdf', 'table' on Weibo @爱可可-爱生活 , getting `camelot` and `py-pdf-parser` for choices.
Frankly speaking,  `camelot` is more famous and has some tutorials on the net, so I try camelot first. Very soon, I meet great difficulties in LOCATING contents in pdf with camelot. Then `py-pdf-parser`does help me, I really appreciate the idea of loacting by wording or font format, it is rather flexible and effective and able to parse contents over pages. I just don't agree with the word on your home page, for the significant advantages: 'When Should I Use Py PDF Parser?'

In conclusion, I guess that a clear and deep tutorial on tech medium or search engine may help popularizing this great work.
Hoping `py-pdf-parser` be best pdf parser!

**Comment by jstockwin** [comment]
Hi @forhonourlx, Thanks for the info - that's cool to know and many thanks for the kind words.

Since `font_mapping_is_regex` has solved your problem, I'll close this issue.

What would you suggest changing the docs to? Maybe I can a new sentence in the first and last paragraphs:

```
Py PDF Parser is for extracting specific, structured data from a PDF. You will be able to write code that should extract for multiple PDFs with the same format. It will also be useful when searching through a PDF for specific bits of data that you'd like to extract.
```

```
If you’re trying to extract specific tables from a certain PDF, other tools (e.g. https://camelot-py.readthedocs.io/en/master/) may be more appropriate, although py-pdf-parser also has some good table extraction tools (link).
```

**Comment by forhonourlx** [comment]
In my opinion, 
```
Py PDF Parser is an excellent tool for locating and extracting specific contents(i.e. texts, tables...) from any comprehensive PDF. You can always find a way to  locate contents you really want: by words or by font format. You may write codes to get flexible manipulation of pdf parse procedure, especially for comprehensive, repeat or over-page tasks.
```

```
Py PDF Parser is good at locating certain tables in comprehensive PDFs, and extracting tables over pages.
If you're trying to extract all tables without locating and over-page problems, other tools
(e.g. https://camelot-py.readthedocs.io/en/master/) may be appropriate in some cases.
```
Forgive my poor English.

**Comment by jstockwin** [comment]
@forhonourlx many thanks! I will use this as a base and update the documentation.

Thanks again for submitting the issue, and I hope you continue to find this project useful!
